pkg/querier/timeline/timeline_test.go

```diff
@@ -82,12 +118,57 @@ func Test_Backfill_All(t *testing.T) {
 		Points: []*typesv1.Point{},
 	}
 
-	timeline := timeline.New(points, startTime, endTime, TimelineStepSec)
+	timeline := timeline.New(points, startTime, endTime, timelineStepSec)
 
 	assert.Equal(t, startTime/1000, timeline.StartTime)
 	assert.Equal(t, []uint64{
-		0, 0, 0, 0, 0,
-		0, 0, 0, 0, 0,
-		0,
+		0, //      0 ms    -60000 ms
+		0, //  10000 ms    -50000 ms
+		0, //  20000 ms    -40000 ms
+		0, //  30000 ms    -30000 ms
+		0, //  40000 ms    -20000 ms
+		0, //  50000 ms    -10000 ms
+		0, //  60000 ms         0 ms (now)
+		0, //  70000 ms    +10000 ms
+		0, //  80000 ms    +20000 ms
+		0, //  90000 ms    +30000 ms
+		0, // 100000 ms    +40000 ms
+		0, // 110000 ms    +50000 ms
 	}, timeline.Samples)
 }
+
+func Test_Backfill_Arbitrary(t *testing.T) {
+	startMs := int64(0)
+	endMs := int64(10 * time.Second / time.Millisecond)
+	step := int64(1)
+	series := &typesv1.Series{
+		Points: []*typesv1.Point{
+			//    0 ms
+			// 1000 ms
+			{Timestamp: 2000, Value: 69},
+			{Timestamp: 3000, Value: 83},
+			// 4000 ms
+			// 5000 ms
+			{Timestamp: 6000, Value: 85},
+			// 7000 ms
+			{Timestamp: 8000, Value: 91},
+			// 9000 ms
+		},
+	}
+
+	tl := timeline.New(series, startMs, endMs, step)
+	assert.Equal(t, startMs/1000, tl.StartTime)
+
+	assert.Equal(t, []uint64{
+		0,  //    0 ms
+		0,  // 1000 ms
+		69, // 2000 ms
+		83, // 3000 ms
+		0,  // 4000 ms
+		0,  // 5000 ms
+		85, // 6000 ms
+		0,  // 7000 ms
+		91, // 8000 ms
+		0,  // 9000 ms
+	}, tl.Samples)
+}
```

Comment:
It looks like we don't add a trailing zero for a sample at endMs

```go
func Test_Backfill_LastSample(t *testing.T) {
	startMs := int64(3000)
	step := int64(10)
	series := &typesv1.Series{
		Points: []*typesv1.Point{
			{Timestamp: 23000, Value: 69},
			{Timestamp: 53000, Value: 91},
		},
	}

	tl := timeline.New(series, startMs, int64(53000), step)

	assert.Equal(t, startMs/1000, tl.StartTime)
	assert.Equal(t, []uint64{
		0, 0, 69, 0, 0, 91,
	}, tl.Samples)

	tl = timeline.New(series, startMs, int64(63000), step)
	assert.Equal(t, startMs/1000, tl.StartTime)
	assert.Equal(t, []uint64{
		0, 0, 69, 0, 0, 91, 0,
	}, tl.Samples)
}

```

Reply:
Is a valid input for `timeline.New` allowed to be equal to `endMs`? The comment for the function seems to indicate this shouldn't be allowed:

> New generates a FlamebearerTimeline,
> backfilling any missing data with zeros
> It assumes:
> * Ordered
> * startMs is earlier than the first series value
> * **endMs is after the last series value**

Though, admittedly the assumptions outlined there seem to suggest `startMs` needs to be strictly before the first series value. I was operating under the assumption series should have timestamps in the range `[startMs, endMs)`.

If this is true, then the 53000ms test case has an invalid series range since it has a timestamp that's not in the range (91). That value has a timestamp equal to `endMs`. While the 63000ms test case has a valid series range, but we shouldn't expect the final zero, since that would correspond to a timestamp outside of `[startMs, endMs)`.

At the end of the day, the logic for how we pad either side of the timeline will depend on what is considered "valid input". If the valid series timestamp range is `[startMs, endMs)`, we'll have the current behavior. If the valid series timestamp range is `[startMs, endMs]`, then we'll have the behavior of this test case above.

This brings up another point, if we are given a series with out of bounds timestamps, should we truncate it to fit within `[startMs, endMs)`?

Reply:
I dont know answers for these questions. I just felt it may be a good idea to keep both cases consistent, either add a zero or drop value. We can merge as is and improve later.